File: ios/StarterApp/Core/Notifications/PushRegistrationManager.swift
import Foundation

@MainActor
final class PushRegistrationManager {
    private let notificationService = NotificationService()

    func registerCurrentDevice(pushToken: String) async throws -> RegisteredPushDevice {
        try await notificationService.registerDevice(token: pushToken)
    }

    func revokeCurrentDevice(pushToken: String) async throws {
        _ = try await notificationService.revokeDevice(token: pushToken)
    }

    func handleRemoteRegistrationSucceeded(deviceToken: Data) async throws -> RegisteredPushDevice {
        let token = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
        return try await registerCurrentDevice(pushToken: token)
    }

    // This manager intentionally does not request APNs permission itself.
    // Permission UX belongs to product UI; registration belongs to starter notification-core.
}

File: android/app/src/main/kotlin/com/appstarterkit/app/core/notifications/PushRegistrationManager.kt
package com.appstarterkit.app.core.notifications

import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class PushRegistrationManager @Inject constructor(
    private val repository: NotificationRepository,
) {
    suspend fun registerCurrentDevice(pushToken: String): Result<RegisteredPushDevice> {
        return repository.registerCurrentDevice(pushToken)
    }

    suspend fun revokeCurrentDevice(pushToken: String): Result<Unit> {
        return repository.revokeCurrentDevice(pushToken)
    }

    // Permission prompts and token retrieval belong to product UI / Firebase setup.
    // Starter notification-core owns the registration contract once a token exists.
}
